### ISSUE TITLE
Add `output_shape` argument to `jax.lax.conv_transpose`.

### ISSUE THREAD

**Issue by hamzamerzic** [google/jax#4592]
TensorFlow v1 and v2 support `output_shape` argument [[link](https://www.tensorflow.org/api_docs/python/tf/nn/conv_transpose)]. Can this be added to JAX?

**Comment by hawkinsp** [comment]
@tamaranorman was looking at this.

**Comment by hamzamerzic** [comment]
We were looking at it together. I'm a bit stuck now, so I'm hoping to get some tips here. Would it be possible to get the same behavior by manipulating user code (e.g. changing `padding`)? If so which, arguments would I need to change. If not, what would we have to expose in `lax.conv_transpose`?

And finally what low-level XLA op is being called by both TF and JAX - that would help in debugging.

**Comment by hawkinsp** [comment]
In all cases the underlying XLA op is `ConvGeneralDilated`. To get a specific output shape from the transposed convolution, you simply need compute the correct padding to get that output shape.

One way to deduce the correct padding would be via `jax.vjp`. A transposed convolution is one of the gradients of a forward convolution, and `jax.vjp` will already know how to compute the correct padding to get back a particular input shape under transposition.

**Comment by hawkinsp** [comment]
@levskaya might be able to help here also.

**Comment by hamzamerzic** [comment]
Thanks for the help Peter. I managed to find the function that would produce the desired before and after padding based on the input. Will include it to Haiku soon. For posterity, this is the function:

```python
def compute_adjusted_padding(input_size: int, output_size: int,
                             kernel_size: int, stride: int, padding: int,
                             dilation: int) -> Tuple[int, int]:
  """Compute adjusted padding."""
  effective_kernel_size = (kernel_size - 1) * dilation + 1
  if padding == "VALID":
    assert input_size == (output_size - effective_kernel_size + stride) // stride
    padding_before = 0
  elif padding == "SAME":
    assert input_size == (output_size + stride - 1) // stride
    padding_needed = (output_size - 1) * stride + effective_kernel_size - output_size
    padding_before = padding_needed // 2

  expanded_input_size = (input_size - 1) * stride + 1
  padded_out_size = output_size + effective_kernel_size - 1
  pad_before = effective_kernel_size - 1 - padding_before
  pad_after = padded_out_size - expanded_input_size - pad_before
  return (pad_before, pad_after)
```

**Comment by hawkinsp** [comment]
Perhaps it would make sense to add `output_shape` support to `lax.conv_transpose`?

**Comment by hamzamerzic** [comment]
Yes I think that would be a great feature. I think that with this function it would be very simple to add. The final bit would be to figure out which input dimensions correspond to which output dimensions. I guess this is stored as part of `ConvDimensionNumbers`, but I have no clue how to interpret those TBH.

**Comment by yang-song** [comment]
+1 to adding this to `lax.conv_transpose`

**Comment by hamzamerzic** [comment]
I've added this capability to Haiku. You can find the computation from `output_shapes` to adjusted padding [here](https://github.com/deepmind/dm-haiku/blob/30b2cb5ce29ea13fc75087146658ec158c007c0a/haiku/_src/conv.py#L571).
